jupyterhub/_version.py

```diff
@@ -42,19 +51,26 @@ def _check_version(hub_version, singleuser_version, log):
         hub_major_minor = V(hub_version).version[:2]
         singleuser_major_minor = V(singleuser_version).version[:2]
         extra = ""
+        do_log = True
         if singleuser_major_minor == hub_major_minor:
             # patch-level mismatch or lower, log difference at debug-level
             # because this should be fine
             log_method = log.debug
         else:
             # log warning-level for more significant mismatch, such as 0.8 vs 0.9, etc.
-            log_method = log.warning
-            extra = " This could cause failure to authenticate and result in redirect loops!"
-        log_method(
-            "jupyterhub version %s != jupyterhub-singleuser version %s." + extra,
-            hub_version,
-            singleuser_version,
-        )
+            global _version_mismatch_warning_logged
+            if _version_mismatch_warning_logged:
+                do_log = False  # We already logged this warning so don't log it again.
+            else:
+                log_method = log.warning
+                extra = " This could cause failure to authenticate and result in redirect loops!"
+                _version_mismatch_warning_logged = True
+        if do_log:
+            log_method(
+                "jupyterhub version %s != jupyterhub-singleuser version %s." + extra,
+                hub_version,
+                singleuser_version,
```

Comment:
Tim mentioned that `singleuser_major_minor` could be different so only logging once per instance of this warning is probably too heavy so the flag should probably be based on `singleuser_major_minor` using a dict.

Reply:
I think the dict is a good idea.

It's probably also good to change this to setting `log_method = log.debug` when it's been seen before, rather than `do_log = False`. Because as it is, if the version doesn't match and it's called for a second time, there's going to be an erroneous debug log saying that the version does match when it doesn't. Logging the mismatch every time at the debug level should still be fine.

Reply:
> Because as it is, if the version doesn't match and it's called for a second time, there's going to be an erroneous debug log saying that the version does match when it doesn't.

I think you might be looking at the nesting wrong here (it's kind of hard to see with the diff in github) but if `do_log` is False we don't log anything. The `else` block below that I think you're looking at is only logged if the versions match which is a different conditional above (L48) and if that is False we don't even have a `do_log` variable.

--

I'm a bit conflicted on what's being asked for direction now. I thought Tim was OK with not repeating the logging for redundant version mismatches as long as I fix this to log each mismatch once. But now it sounds like you (Min) would like me to log everything where the first mismatch for a particular version combination is at `warning` level and any redundancies thereafter are logged at `debug`, but I don't see what the redundant `debug` logging buys us.